rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,565 @@
+a warning will be emitted and the initial generation of the BotInstance will
+be sourced from the certificates current generation counter. This behaviour
+will support the migration of existing `tbot` instances to the new BotInstance
+behaviour.
+
+#### Recording Heartbeat Data
+
+A new RPC will be added for submitting heartbeat data:
+
+```protobuf
+syntax = "proto3";
+
+package teleport.machineid.v1;
+
+service BotInstanceService {
+  // SubmitHeartbeat submits a heartbeat for a BotInstance.
+  rpc SubmitHeartbeat(SubmitHeartbeatRequest) returns (SubmitHeartbeatResponse);
+}
+
+// The request for SubmitHeartbeat.
+message SubmitHeartbeatRequest {
+  // The heartbeat data to submit.
+  BotInstanceStatusHeartbeat heartbeat = 1;
+}
+
+// The response for SubmitHeartbeat.
+message SubmitHeartbeatResponse {
+  // Empty
+}
+```
+
+The endpoint will have a special auth/authz check. RBAC will not be used and
+instead the endpoint will check:
+
+- The presented client certificate is for the Bot linked to the instance.
+- The presented client certificate's public key matches the public key recorded
+  for the BotInstance.
+
+This endpoint will be called by `tbot` immediately after it has initially
+authenticated. After a heartbeat has successfully completed, another should be
+scheduled for a half hour after. A small amount of jitter should be added to
+the heartbeat period to avoid a thundering herd of heartbeats.
+
+If the heartbeat fails, then `tbot` should retry on an exponential backoff.
+
+##### Alternative: Submit Heartbeat Data on Join/Renew
+
+Alternatively, we could add a Heartbeat field to the join/renew RPCs.
+
+Pros:
+
+- Avoids introducing a new RPC and ensures that all data within the Heartbeat
+  comes from the same instance in time.
+- Allows self-reported information to be used as part of renewal decision.
+  This is not a strong defence as it is self-reported and cannot be trusted.
+- Avoids a state where the BotInstance is incomplete immediately after joining
+  and before it has called SubmitHeartbeat.
+
+Cons:
+
+- Adds Bot specific behaviour to RPCs that are also used for Node joining.
+- Heartbeats are limited to the interval of renewal.
+
+#### API
+
+Additional RPCs will be added to the BotInstance service to allow these to
+be listed and deleted:
+
+```protobuf
+syntax = "proto3";
+
+package teleport.machineid.v1;
+
+service BotInstanceService {
+  // GetBotInstance returns the specified BotInstance resource.
+  rpc GetBotInstance(GetBotInstanceRequest) returns (BotInstance);
+  // ListBotInstances returns a page of BotInstance resources.
+  rpc ListBotInstances(ListBotInstancesRequest) returns (ListBotInstancesResponse);
+  // DeleteBotInstance hard deletes the specified BotInstance resource.
+  rpc DeleteBotInstance(DeleteBotInstanceRequest) returns (google.protobuf.Empty);
+}
+
+// Request for GetBotInstance.
+message GetBotInstanceRequest {
+  // The name of the BotInstance to retrieve.
+  string name = 1;
```

Comment:
I wonder if it makes more sense to seperate the bot name and fingerprint fields here rather than provide them concatenated.

Reply:
Yeah -  I think I'm growing to like the idea of explicit fields - but I do see one potential issue. If I already have a BotInstance resource, it's very easy to call GetBotInstance using the name of the resource in this field if I want to  check for any updates. If we switch to separate fields, there's potentially a little more complexity in this as you'd have to do something like

```
GetBotInstance({
  Bot: instance.spec.bot,
  Instance: instance.name
})
```

This feels fairly minor so I haven't got a super strong opinion here. Feel free to change.

Reply:
I've split the fields, though on some level I wonder if we could just refer directly to the bot instance identifier. The bot name namespaces the UUID, but is that really necessary?

Reply:
Hm - but with only the UUID, we won't be able to get the BotInstance from the store, as we're going to include the bot name in the storage key to enable listing of BotInstances for a given bot.

Reply:
Ah, good call, didn't think about that.